tests/web/test_db_query.py

```diff
@@ -478,6 +482,132 @@ def test_db_query_fail() -> None:
     assert excinfo.value.extra["sql"] is not None
 
 
+def test_apply_allocation_policies_quota_sets_throttle_policy() -> None:
+    query, storage, _ = _build_test_query("count(distinct(project_id))")
+    update_called = False
+
+    class ThrottleAllocationPolicy1(AllocationPolicy):
+        def _additional_config_definitions(self) -> list[AllocationPolicyConfig]:
+            return []
+
+        def _get_quota_allowance(
+            self, tenant_ids: dict[str, str | int], query_id: str
+        ) -> QuotaAllowance:
+            return QuotaAllowance(
+                can_run=True,
+                max_threads=1,
+                explanation={
+                    "reason": "ThrottleAllocationPolicy1 throttles all queries"
+                },
+                is_throttled=True,
+                throttle_threshold=MAX_THRESHOLD,
+                rejection_threshold=MAX_THRESHOLD + 1,
+                quota_used=MAX_THRESHOLD,
+                quota_unit=NO_UNITS,
+                suggestion=NO_SUGGESTION,
+            )
+
+        def _update_quota_balance(
+            self,
+            tenant_ids: dict[str, str | int],
+            query_id: str,
+            result_or_error: QueryResultOrError,
+        ) -> None:
+            nonlocal update_called
+            update_called = True
+            return
+
+    class ThrottleAllocationPolicy2(AllocationPolicy):
+        def _additional_config_definitions(self) -> list[AllocationPolicyConfig]:
+            return []
+
+        def _get_quota_allowance(
+            self, tenant_ids: dict[str, str | int], query_id: str
+        ) -> QuotaAllowance:
+            return QuotaAllowance(
+                can_run=True,
+                max_threads=2,
+                explanation={
+                    "reason": "ThrottleAllocationPolicy2 throttles all queries"
+                },
+                is_throttled=True,
+                throttle_threshold=MAX_THRESHOLD,
+                rejection_threshold=MAX_THRESHOLD + 1,
+                quota_used=MAX_THRESHOLD,
+                quota_unit=NO_UNITS,
+                suggestion=NO_SUGGESTION,
+            )
+
+        def _update_quota_balance(
+            self,
+            tenant_ids: dict[str, str | int],
+            query_id: str,
+            result_or_error: QueryResultOrError,
+        ) -> None:
+            nonlocal update_called
+            update_called = True
+            return
```

Comment:
would it be a little more readable if the output was an argument to the constructor here and you just had 1 class (with two instances)? Or does the logic require them to actually be two separate classes

Reply:
What do you mean by "output was an argument to the constructor"?

The logic I'm testing here is to have 2 different policies. In reality they could be `ReferrerGuardRailPolicy` and `BytesScannedRejectingPolicy`, which are separate classes.

Reply:
I'm imagining there's a class called `MockAllocationPolicy` and it looks something like

```python
class MockAllocationPolicy(AllocationPolicy):
   def __init__(quota_allowance_value):
      self.__quota_allowance = quota_allowance_value
  
   def _get_quota_allowance(self, tenant_ids: dict[str, str | int], query_id: str):
     return self.__quota_allowance
```

and you instantiate two instances of that rather than having two subclasses that do essentially the same thing with a slightly different faked return.

That said, looking at the code a bit further, I think the class names do have semantic value here and you really do need two separate subclasses.

Reply:
We could have `MockAllocationPolicy` inherit from `AllocationPolicy` and then `MockAllocationPolicy` implement the currently duplicate functionality. Then inherit `ThrottleAllocationPolicy1` and `ThrottleAllocationPolicy2` from `MockAllocationPolicy` and override `_get_quota_allowance` for things like `max_threads` in each Throttle class. That should reduce duplication, increase readability and still have 2 classes like you want.

Reply:
Thank you Onkar and Oliver! I've made the changes.